docs/code_from_chapters/usecase_ml_code.rst

```diff
@@ -0,0 +1,840 @@
+
+   # show the directory hierarchy
+   tree -d
+
+In order to link the correct software environment to the data we add a prepared software container with the correct Python libraries for the analysis::
+
+   datalad containers-add software --url shub://adswa/python-ml:1
+
+Prepare the data
+""""""""""""""""
+
+This workflow uses a script to create labeled sets of training and validation data -- for the sake of this example, only two categories or ten are labeled.
+The script will read out file names from the input data, and create csv files that map file names to image categories::
+
+   cat << EOT > code/prepare.py
+   #!/usr/bin/env python3
+
+   import pandas as pd
+   from pathlib import Path
+
+   FOLDERS_TO_LABELS = {"n03445777": "golf ball",
+                        "n03888257": "parachute"}
+
+
+   def get_files_and_labels(source_path):
+       images = []
+       labels = []
+       for image_path in source_path.rglob("*/*.JPEG"):
+           filename = image_path
+           folder = image_path.parent.name
+           if folder in FOLDERS_TO_LABELS:
+               images.append(filename)
+               label = FOLDERS_TO_LABELS[folder]
+               labels.append(label)
+       return images, labels
+
+
+   def save_as_csv(filenames, labels, destination):
+       data_dictionary = {"filename": filenames, "label": labels}
+       data_frame = pd.DataFrame(data_dictionary)
+       data_frame.to_csv(destination)
+
+
+   def main(repo_path):
+       data_path = repo_path / "data"
+       train_path = data_path / "raw/imagenette2-160/train"
+       test_path = data_path / "raw/imagenette2-160/val"
+       train_files, train_labels = get_files_and_labels(train_path)
+       test_files, test_labels = get_files_and_labels(test_path)
+       save_as_csv(train_files, train_labels, data_path / "train.csv")
+       save_as_csv(test_files, test_labels, data_path / "test.csv")
+
+
+   if __name__ == "__main__":
+       repo_path = Path(__file__).parent.parent
+       main(repo_path)
+   EOT
+
+This yields a new, untracked file::
+
+   datalad status
+
+And we can save it, optionally with a version tag::
+
+   datalad save -m "Add script for data preparation for 2 categories" \
+      --version-tag "ready4prepping" code/prepare.py
+
+
+We prepare the data using :command:`datalad containers-run` to ensure that all relevant Python libraries are installed.
+To keep execution time in this example short, we only specify the subset of the data that the above script uses as an input::
+
+
+   datalad containers-run -n software \
+     -m "Prepare the data for categories golf balls and parachutes" \
+     --input 'data/raw/imagenette2-160/train/n03445777' \
+     --input 'data/raw/imagenette2-160/val/n03445777' \
+     --input 'data/raw/imagenette2-160/train/n03888257' \
+     --input 'data/raw/imagenette2-160/val/n03888257' \
+     --output 'data/train.csv' \
+     --output 'data/test.csv' \
+     "python3 code/prepare.py"
```

Comment:
Download can be sped up with

`datalad -c datalad.runtime.max-annex-jobs=5 containers-run ...`

Reply:
thx